File: src/main/java/no/fintlabs/flyt/gateway/application/archive/resource/kodeverk/CodelistController.java
package no.fintlabs.flyt.gateway.application.archive.resource.kodeverk;

import no.fint.model.felles.basisklasser.Begrep;
import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.arkiv.kodeverk.*;
import no.fint.model.resource.arkiv.noark.AdministrativEnhetResource;
import no.fint.model.resource.arkiv.noark.ArkivdelResource;
import no.fint.model.resource.arkiv.noark.ArkivressursResource;
import no.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource;
import no.fintlabs.cache.FintCache;
import no.fintlabs.flyt.gateway.application.archive.links.ResourceLinkUtil;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Collection;
import java.util.StringJoiner;
import java.util.stream.Collectors;

import static no.fintlabs.resourceserver.UrlPaths.INTERNAL_API;

@RestController
@RequestMapping(INTERNAL_API + "/arkiv/kodeverk")
public class CodelistController {

    private final FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache;
    private final FintCache<String, ArkivdelResource> arkivdelResourceCache;
    private final FintCache<String, ArkivressursResource> arkivressursResourceCache;
    private final FintCache<String, TilknyttetRegistreringSomResource> tilknyttetRegistreringSomResourceCache;
    private final FintCache<String, DokumentStatusResource> dokumentStatusResourceCache;
    private final FintCache<String, DokumentTypeResource> dokumentTypeResourceCache;
    private final FintCache<String, KlassifikasjonssystemResource> klassifikasjonssystemResourceCache;
    private final FintCache<String, PartRolleResource> partRolleResourceCache;
    private final FintCache<String, KorrespondansepartTypeResource> korrespondansepartTypeResourceCache;
    private final FintCache<String, SaksstatusResource> saksstatusResourceCache;
    private final FintCache<String, SkjermingshjemmelResource> skjermingshjemmelResourceCache;
    private final FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache;
    private final FintCache<String, JournalStatusResource> journalStatusResourceCache;
    private final FintCache<String, JournalpostTypeResource> journalpostTypeResourceCache;
    private final FintCache<String, SaksmappetypeResource> saksmappetypeResourceCache;
    private final FintCache<String, VariantformatResource> variantformatResourceCache;
    private final FintCache<String, FormatResource> formatResourceCache;
    private final FintCache<String, TilgangsgruppeResource> tilgangsgruppeResourceCache;
    private final ArkivressursDisplayNameMapper arkivressursDisplayNameMapper;

    public CodelistController(
            FintCache<String, AdministrativEnhetResource> administrativEnhetResourceCache,
            FintCache<String, ArkivdelResource> arkivdelResourceCache,
            FintCache<String, ArkivressursResource> arkivressursResourceCache,
            FintCache<String, TilknyttetRegistreringSomResource> tilknyttetRegistreringSomResourceCache,
            FintCache<String, DokumentStatusResource> dokumentStatusResourceCache,
            FintCache<String, DokumentTypeResource> dokumentTypeResourceCache,
            FintCache<String, KlassifikasjonssystemResource> klassifikasjonssystemResourceCache,
            FintCache<String, PartRolleResource> partRolleResourceCache,
            FintCache<String, KorrespondansepartTypeResource> korrespondansepartTypeResourceCache,
            FintCache<String, SaksstatusResource> saksstatusResourceCache,
            FintCache<String, SkjermingshjemmelResource> skjermingshjemmelResourceCache,
            FintCache<String, TilgangsrestriksjonResource> tilgangsrestriksjonResourceCache,
            FintCache<String, JournalStatusResource> journalStatusResourceCache,
            FintCache<String, JournalpostTypeResource> journalpostTypeResourceCache,
            FintCache<String, SaksmappetypeResource> saksmappetypeResourceCache,
            FintCache<String, VariantformatResource> variantformatResourceCache,
            FintCache<String, FormatResource> formatResourceCache,
            FintCache<String, TilgangsgruppeResource> tilgangsgruppeResourceCache,
            ArkivressursDisplayNameMapper arkivressursDisplayNameMapper
    ) {
        this.administrativEnhetResourceCache = administrativEnhetResourceCache;
        this.arkivdelResourceCache = arkivdelResourceCache;
        this.arkivressursResourceCache = arkivressursResourceCache;
        this.tilknyttetRegistreringSomResourceCache = tilknyttetRegistreringSomResourceCache;
        this.dokumentStatusResourceCache = dokumentStatusResourceCache;
        this.dokumentTypeResourceCache = dokumentTypeResourceCache;
        this.klassifikasjonssystemResourceCache = klassifikasjonssystemResourceCache;
        this.partRolleResourceCache = partRolleResourceCache;
        this.korrespondansepartTypeResourceCache = korrespondansepartTypeResourceCache;
        this.saksstatusResourceCache = saksstatusResourceCache;
        this.skjermingshjemmelResourceCache = skjermingshjemmelResourceCache;
        this.tilgangsrestriksjonResourceCache = tilgangsrestriksjonResourceCache;
        this.journalStatusResourceCache = journalStatusResourceCache;
        this.journalpostTypeResourceCache = journalpostTypeResourceCache;
        this.saksmappetypeResourceCache = saksmappetypeResourceCache;
        this.variantformatResourceCache = variantformatResourceCache;
        this.formatResourceCache = formatResourceCache;
        this.tilgangsgruppeResourceCache = tilgangsgruppeResourceCache;
        this.arkivressursDisplayNameMapper = arkivressursDisplayNameMapper;
    }

    @GetMapping("administrativenhet")
    public ResponseEntity<Collection<ResourceReference>> getAdministrativEnheter() {
        return ResponseEntity.ok(
                administrativEnhetResourceCache
                        .getAllDistinct()
                        .stream()
                        .map(administrativEnhetResource -> this.mapToResourceReference(
                                administrativEnhetResource,
                                new ResourceReferenceDisplayNameBuilder()
                                        .technicalId(administrativEnhetResource.getSystemId())
                                        .name(administrativEnhetResource.getNavn())
                        ))
                        .collect(Collectors.toList())
        );
    }

    @GetMapping("klassifikasjonssystem")
    public ResponseEntity<Collection<ResourceReference>> getKlassifikasjonssystem() {
        return ResponseEntity.ok(
                klassifikasjonssystemResourceCache
                        .getAllDistinct()
                        .stream()
                        .map(klassifikasjonssystemResource -> this.mapToResourceReference(
                                klassifikasjonssystemResource,
                                new ResourceReferenceDisplayNameBuilder()
                                        .technicalId(klassifikasjonssystemResource.getSystemId())
                                        .name(klassifikasjonssystemResource.getTittel())
                        ))
                        .collect(Collectors.toList())
        );
    }

    @GetMapping("klasse")
    public ResponseEntity<Collection<ResourceReference>> getKlasse(@RequestParam String klassifikasjonssystemLink) {
        return klassifikasjonssystemResourceCache.getOptional(klassifikasjonssystemLink)
                .map(KlassifikasjonssystemResource::getKlasse)
                .<Collection<ResourceReference>>map(klasseResources -> klasseResources.stream()
                        .map(klasse -> this.mapToResourceReference(
                                klasse.getKlasseId(),
                                new ResourceReferenceDisplayNameBuilder()
                                        .functionalId(klasse.getKlasseId())
                                        .name(klasse.getTittel())
                        )).toList()
                )
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("arkivdel")
    public ResponseEntity<Collection<ResourceReference>> getArkivdel() {
        return ResponseEntity.ok(
                arkivdelResourceCache
                        .getAllDistinct()
                        .stream()
                        .map(arkivdelResource -> this.mapToResourceReference(
                                arkivdelResource,
                                new ResourceReferenceDisplayNameBuilder()
                                        .technicalId(arkivdelResource.getSystemId())
                                        .name(arkivdelResource.getTittel())
                        ))
                        .collect(Collectors.toList())
        );
    }

    @GetMapping("arkivressurs")
    public ResponseEntity<Collection<ResourceReference>> getArkivressurs() {
        return ResponseEntity.ok(
                arkivressursResourceCache
                        .getAllDistinct()
                        .stream()
                        .map(arkivressurs -> {
                            ResourceReferenceDisplayNameBuilder resourceReferenceDisplayNameBuilder = new ResourceReferenceDisplayNameBuilder()
                                    .technicalId(arkivressurs.getSystemId());

                            arkivressursDisplayNameMapper.findPersonalressursBrukernavn(arkivressurs)
                                    .ifPresent(resourceReferenceDisplayNameBuilder::functionalId);

                            arkivressursDisplayNameMapper.findPersonNavn(arkivressurs)
                                    .ifPresent(resourceReferenceDisplayNameBuilder::name);

                            return mapToResourceReference(arkivressurs, resourceReferenceDisplayNameBuilder);
                        })
                        .collect(Collectors.toList())
        );
    }

    @GetMapping("partrolle")
    public ResponseEntity<Collection<ResourceReference>> getPartRolle() {
        return getBegrepResourceReferences(partRolleResourceCache);
    }

    @GetMapping("korrespondanseparttype")
    public ResponseEntity<Collection<ResourceReference>> getKorrespondansepartType() {
        return getBegrepResourceReferences(korrespondansepartTypeResourceCache);
    }

    @GetMapping("tilknyttetregistreringsom")
    public ResponseEntity<Collection<ResourceReference>> getTilknyttetRegistreringSom() {
        return getBegrepResourceReferences(tilknyttetRegistreringSomResourceCache);
    }

    @GetMapping("sakstatus")
    public ResponseEntity<Collection<ResourceReference>> getSakstatus() {
        return getBegrepResourceReferences(saksstatusResourceCache);
    }

    @GetMapping("skjermingshjemmel")
    public ResponseEntity<Collection<ResourceReference>> getSkjermingshjemmel() {
        return getBegrepResourceReferences(skjermingshjemmelResourceCache);
    }

    @GetMapping("tilgangsrestriksjon")
    public ResponseEntity<Collection<ResourceReference>> getTilgangsrestriksjon() {
        return getBegrepResourceReferences(tilgangsrestriksjonResourceCache);
    }

    @GetMapping("dokumentstatus")
    public ResponseEntity<Collection<ResourceReference>> getDokumentstatus() {
        return getBegrepResourceReferences(dokumentStatusResourceCache);
    }

    @GetMapping("dokumenttype")
    public ResponseEntity<Collection<ResourceReference>> getDokumenttype() {
        return getBegrepResourceReferences(dokumentTypeResourceCache);
    }

    @GetMapping("journalstatus")
    public ResponseEntity<Collection<ResourceReference>> getJournalstatus() {
        return getBegrepResourceReferences(journalStatusResourceCache);
    }

    @GetMapping("journalposttype")
    public ResponseEntity<Collection<ResourceReference>> getJournalposttype() {
        return getBegrepResourceReferences(journalpostTypeResourceCache);
    }

    @GetMapping("saksmappetype")
    public ResponseEntity<Collection<ResourceReference>> getSaksmappetype() {
        return getBegrepResourceReferences(saksmappetypeResourceCache);
    }

    @GetMapping("variantformat")
    public ResponseEntity<Collection<ResourceReference>> getVariantformat() {
        return getBegrepResourceReferences(variantformatResourceCache);
    }

    @GetMapping("format")
    public ResponseEntity<Collection<ResourceReference>> getFormat() {
        return getBegrepResourceReferences(formatResourceCache);
    }

    @GetMapping("tilgangsgruppe")
    public ResponseEntity<Collection<ResourceReference>> getTilgangsgruppe() {
        return getBegrepResourceReferences(tilgangsgruppeResourceCache);
    }

    private <R extends Begrep & FintLinks> ResponseEntity<Collection<ResourceReference>> getBegrepResourceReferences(FintCache<String, R> resouceCache) {
        return ResponseEntity.ok(
                resouceCache
                        .getAllDistinct()
                        .stream()
                        .map(this::mapToResourceReference)
                        .collect(Collectors.toList())
        );
    }

    private <R extends Begrep & FintLinks> ResourceReference mapToResourceReference(R resource) {
        return new ResourceReference(
                ResourceLinkUtil.getFirstSelfLink(resource),
                new ResourceReferenceDisplayNameBuilder()
                        .functionalId(resource.getKode())
                        .technicalId(resource.getSystemId())
                        .name(resource.getNavn())
                        .build()
        );
    }

    private ResourceReference mapToResourceReference(FintLinks resource, ResourceReferenceDisplayNameBuilder resourceReferenceDisplayNameBuilder) {
        return new ResourceReference(
                ResourceLinkUtil.getFirstSelfLink(resource),
                resourceReferenceDisplayNameBuilder.build()
        );
    }

    private ResourceReference mapToResourceReference(String id, ResourceReferenceDisplayNameBuilder resourceReferenceDisplayNameBuilder) {
        return new ResourceReference(
                id,
                resourceReferenceDisplayNameBuilder.build()
        );
    }

    private static class ResourceReferenceDisplayNameBuilder {
        private String functionalId;
        private String technicalId;
        private String name;

        ResourceReferenceDisplayNameBuilder functionalId(String functionalId) {
            this.functionalId = functionalId;
            return this;
        }

        ResourceReferenceDisplayNameBuilder technicalId(Identifikator technicalId) {
            return technicalId(technicalId.getIdentifikatorverdi());
        }

        ResourceReferenceDisplayNameBuilder technicalId(String technicalId) {
            this.technicalId = technicalId;
            return this;
        }

        ResourceReferenceDisplayNameBuilder name(String name) {
            this.name = name;
            return this;
        }

        String build() {
            StringJoiner stringJoiner = new StringJoiner(" ");
            if (functionalId != null) {
                stringJoiner.add("[" + functionalId + "]");
            }
            if (name != null) {
                stringJoiner.add(name);
            }
            if (technicalId != null) {
                stringJoiner.add("#" + technicalId);
            }
            return stringJoiner.toString();
        }

    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/resource/kodeverk/CodelistController.kt
package no.novari.flyt.archive.gateway.resource.kodeverk

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import no.novari.cache.FintCache
import no.novari.fint.model.felles.basisklasser.Begrep
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.arkiv.kodeverk.DokumentStatusResource
import no.novari.fint.model.resource.arkiv.kodeverk.DokumentTypeResource
import no.novari.fint.model.resource.arkiv.kodeverk.FormatResource
import no.novari.fint.model.resource.arkiv.kodeverk.JournalStatusResource
import no.novari.fint.model.resource.arkiv.kodeverk.JournalpostTypeResource
import no.novari.fint.model.resource.arkiv.kodeverk.KorrespondansepartTypeResource
import no.novari.fint.model.resource.arkiv.kodeverk.PartRolleResource
import no.novari.fint.model.resource.arkiv.kodeverk.SaksmappetypeResource
import no.novari.fint.model.resource.arkiv.kodeverk.SaksstatusResource
import no.novari.fint.model.resource.arkiv.kodeverk.SkjermingshjemmelResource
import no.novari.fint.model.resource.arkiv.kodeverk.TilgangsgruppeResource
import no.novari.fint.model.resource.arkiv.kodeverk.TilgangsrestriksjonResource
import no.novari.fint.model.resource.arkiv.kodeverk.TilknyttetRegistreringSomResource
import no.novari.fint.model.resource.arkiv.kodeverk.VariantformatResource
import no.novari.fint.model.resource.arkiv.noark.AdministrativEnhetResource
import no.novari.fint.model.resource.arkiv.noark.ArkivdelResource
import no.novari.fint.model.resource.arkiv.noark.ArkivressursResource
import no.novari.fint.model.resource.arkiv.noark.KlasseResource
import no.novari.fint.model.resource.arkiv.noark.KlassifikasjonssystemResource
import no.novari.flyt.archive.gateway.links.ResourceLinkUtil
import no.novari.flyt.webresourceserver.UrlPaths.INTERNAL_API
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("$INTERNAL_API/arkiv/kodeverk")
@Tag(name = "Kodeverk", description = "Cachede arkivkodeverk og arkivressurser.")
class CodelistController(
    private val administrativEnhetResourceCache: FintCache<String, AdministrativEnhetResource>,
    private val arkivdelResourceCache: FintCache<String, ArkivdelResource>,
    private val arkivressursResourceCache: FintCache<String, ArkivressursResource>,
    private val tilknyttetRegistreringSomResourceCache: FintCache<String, TilknyttetRegistreringSomResource>,
    private val dokumentStatusResourceCache: FintCache<String, DokumentStatusResource>,
    private val dokumentTypeResourceCache: FintCache<String, DokumentTypeResource>,
    private val klassifikasjonssystemResourceCache: FintCache<String, KlassifikasjonssystemResource>,
    private val partRolleResourceCache: FintCache<String, PartRolleResource>,
    private val korrespondansepartTypeResourceCache: FintCache<String, KorrespondansepartTypeResource>,
    private val saksstatusResourceCache: FintCache<String, SaksstatusResource>,
    private val skjermingshjemmelResourceCache: FintCache<String, SkjermingshjemmelResource>,
    private val tilgangsrestriksjonResourceCache: FintCache<String, TilgangsrestriksjonResource>,
    private val journalStatusResourceCache: FintCache<String, JournalStatusResource>,
    private val journalpostTypeResourceCache: FintCache<String, JournalpostTypeResource>,
    private val saksmappetypeResourceCache: FintCache<String, SaksmappetypeResource>,
    private val variantformatResourceCache: FintCache<String, VariantformatResource>,
    private val formatResourceCache: FintCache<String, FormatResource>,
    private val tilgangsgruppeResourceCache: FintCache<String, TilgangsgruppeResource>,
    private val arkivressursDisplayNameMapper: ArkivressursDisplayNameMapper,
) {
    @GetMapping("administrativenhet")
    @Operation(summary = "List administrative enheter")
    fun getAdministrativEnheter(): ResponseEntity<Collection<ResourceReference>> =
        ResponseEntity.ok(
            administrativEnhetResourceCache
                .getAllDistinct()
                .map { administrativEnhetResource ->
                    mapToResourceReference(
                        administrativEnhetResource,
                        ResourceReferenceDisplayNameBuilder()
                            .technicalId(administrativEnhetResource.systemId)
                            .name(administrativEnhetResource.navn),
                    )
                },
        )

    @GetMapping("klassifikasjonssystem")
    @Operation(summary = "List klassifikasjonssystemer")
    fun getKlassifikasjonssystem(): ResponseEntity<Collection<ResourceReference>> =
        ResponseEntity.ok(
            klassifikasjonssystemResourceCache
                .getAllDistinct()
                .map { klassifikasjonssystemResource ->
                    mapToResourceReference(
                        klassifikasjonssystemResource,
                        ResourceReferenceDisplayNameBuilder()
                            .technicalId(klassifikasjonssystemResource.systemId)
                            .name(klassifikasjonssystemResource.tittel),
                    )
                },
        )

    @GetMapping("klasse")
    @Operation(
        summary = "List klasser for klassifikasjonssystem",
        description = "Returnerer klasser som hører til et gitt klassifikasjonssystem.",
    )
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Klasser funnet."),
            ApiResponse(
                responseCode = "404",
                description = "Fant ikke klassifikasjonssystemet eller tilhørende klasser.",
            ),
        ],
    )
    fun getKlasse(
        @Parameter(description = "Self-link for klassifikasjonssystemet.", required = true)
        @RequestParam klassifikasjonssystemLink: String,
    ): ResponseEntity<Collection<ResourceReference>> {
        val klasseResources: List<KlasseResource> =
            klassifikasjonssystemResourceCache
                .getOptional(klassifikasjonssystemLink)
                .orElse(null)
                ?.klasse
                ?.toList()
                ?: emptyList()
        val resourceReferences =
            klasseResources.map { klasse ->
                mapToResourceReference(
                    klasse.klasseId,
                    ResourceReferenceDisplayNameBuilder()
                        .functionalId(klasse.klasseId)
                        .name(klasse.tittel),
                )
            }

        return if (resourceReferences.isNotEmpty()) {
            ResponseEntity.ok(resourceReferences)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    @GetMapping("arkivdel")
    @Operation(summary = "List arkivdeler")
    fun getArkivdel(): ResponseEntity<Collection<ResourceReference>> =
        ResponseEntity.ok(
            arkivdelResourceCache
                .getAllDistinct()
                .map { arkivdelResource ->
                    mapToResourceReference(
                        arkivdelResource,
                        ResourceReferenceDisplayNameBuilder()
                            .technicalId(arkivdelResource.systemId)
                            .name(arkivdelResource.tittel),
                    )
                },
        )

    @GetMapping("arkivressurs")
    @Operation(summary = "List arkivressurser")
    fun getArkivressurs(): ResponseEntity<Collection<ResourceReference>> =
        ResponseEntity.ok(
            arkivressursResourceCache
                .getAllDistinct()
                .map { arkivressurs ->
                    val functionalId =
                        arkivressursDisplayNameMapper
                            .findPersonalressursBrukernavn(arkivressurs)
                            .orElse(null)
                    val name =
                        arkivressursDisplayNameMapper
                            .findPersonNavn(arkivressurs)
                            .orElse(null)
                    val technicalId = arkivressurs.systemId?.identifikatorverdi
                    val displayNameBuilder =
                        ResourceReferenceDisplayNameBuilder()
                            .technicalId(technicalId)
                            .functionalId(functionalId)
                            .name(name)

                    ResourceReference(
                        id = ResourceLinkUtil.getFirstSelfLink(arkivressurs),
                        displayName = displayNameBuilder.build(),
                        functionalId = functionalId.orEmpty(),
                        name = name.orEmpty(),
                        technicalId = technicalId.orEmpty(),
                    )
                },
        )

    @GetMapping("partrolle")
    @Operation(summary = "List partroller")
    fun getPartRolle(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(partRolleResourceCache)

    @GetMapping("korrespondanseparttype")
    @Operation(summary = "List korrespondanseparttyper")
    fun getKorrespondansepartType(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(korrespondansepartTypeResourceCache)

    @GetMapping("tilknyttetregistreringsom")
    @Operation(summary = "List tilknyttet registrering som")
    fun getTilknyttetRegistreringSom(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(tilknyttetRegistreringSomResourceCache)

    @GetMapping("sakstatus")
    @Operation(summary = "List saksstatuser")
    fun getSakstatus(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(saksstatusResourceCache)

    @GetMapping("skjermingshjemmel")
    @Operation(summary = "List skjermingshjemler")
    fun getSkjermingshjemmel(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(skjermingshjemmelResourceCache)

    @GetMapping("tilgangsrestriksjon")
    @Operation(summary = "List tilgangsrestriksjoner")
    fun getTilgangsrestriksjon(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(tilgangsrestriksjonResourceCache)

    @GetMapping("dokumentstatus")
    @Operation(summary = "List dokumentstatuser")
    fun getDokumentstatus(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(dokumentStatusResourceCache)

    @GetMapping("dokumenttype")
    @Operation(summary = "List dokumenttyper")
    fun getDokumenttype(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(dokumentTypeResourceCache)

    @GetMapping("journalstatus")
    @Operation(summary = "List journalstatuser")
    fun getJournalstatus(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(journalStatusResourceCache)

    @GetMapping("journalposttype")
    @Operation(summary = "List journalposttyper")
    fun getJournalposttype(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(journalpostTypeResourceCache)

    @GetMapping("saksmappetype")
    @Operation(summary = "List saksmappetyper")
    fun getSaksmappetype(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(saksmappetypeResourceCache)

    @GetMapping("variantformat")
    @Operation(summary = "List variantformater")
    fun getVariantformat(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(variantformatResourceCache)

    @GetMapping("format")
    @Operation(summary = "List formater")
    fun getFormat(): ResponseEntity<Collection<ResourceReference>> = getBegrepResourceReferences(formatResourceCache)

    @GetMapping("tilgangsgruppe")
    @Operation(summary = "List tilgangsgrupper")
    fun getTilgangsgruppe(): ResponseEntity<Collection<ResourceReference>> =
        getBegrepResourceReferences(tilgangsgruppeResourceCache)

    private fun <R> getBegrepResourceReferences(
        resourceCache: FintCache<String, R>,
    ): ResponseEntity<Collection<ResourceReference>>
        where R : Begrep,
              R : FintLinks =
        ResponseEntity.ok(resourceCache.getAllDistinct().map(this::mapToResourceReference))

    private fun <R> mapToResourceReference(
        resource: R,
    ): ResourceReference
        where R : Begrep,
              R : FintLinks =
        ResourceReference(
            ResourceLinkUtil.getFirstSelfLink(resource),
            ResourceReferenceDisplayNameBuilder()
                .functionalId(resource.kode)
                .technicalId(resource.systemId)
                .name(resource.navn)
                .build(),
        )

    private fun mapToResourceReference(
        resource: FintLinks,
        resourceReferenceDisplayNameBuilder: ResourceReferenceDisplayNameBuilder,
    ): ResourceReference =
        ResourceReference(ResourceLinkUtil.getFirstSelfLink(resource), resourceReferenceDisplayNameBuilder.build())

    private fun mapToResourceReference(
        id: String,
        resourceReferenceDisplayNameBuilder: ResourceReferenceDisplayNameBuilder,
    ): ResourceReference = ResourceReference(id, resourceReferenceDisplayNameBuilder.build())

    private class ResourceReferenceDisplayNameBuilder {
        private var functionalId: String? = null
        private var technicalId: String? = null
        private var name: String? = null

        fun functionalId(functionalId: String?) = apply { this.functionalId = functionalId }

        fun technicalId(technicalId: Identifikator?) = apply { this.technicalId = technicalId?.identifikatorverdi }

        fun technicalId(technicalId: String?) = apply { this.technicalId = technicalId }

        fun name(name: String?) = apply { this.name = name }

        fun build(): String =
            buildList {
                functionalId?.let { add("[$it]") }
                name?.let(::add)
                technicalId?.let { add("#$it") }
            }.joinToString(" ")
    }
}
